### PR TITLE
Snapshot Testing

### PR DESCRIPTION
**Pull request by ShaunLawrie**
This uses namespace imports for spectre.console consistently in the main module code. I'm not sure I'm a fan of using `using` statements because it makes it harder to copy paste parts into the terminal for testing but at least it's the same in all public functions now.

This switches to using the `Spectre.Console.Testing.TestConsole` to allow easy inspection of "terminal" output during tests. This allows suppressing output of the test run and evaluating it afterwards.

For most visual outputs I've added snapshot tests that look like this when they fail:
image

### CODE REVIEW

**Comment by ShaunLawrie** [comment]
@trackd I hope I didn't break your tests/code switching them to the testconsole. It's part of the spectre.console.testing library you linked a while ago. The testconsole is pretty good, you can set it up and specify things like the width, features like ansi etc. so the tests are easier to get to work locally + in gh actions

**Comment by trackd** [comment]
> This uses namespace imports for spectre.console consistently in the main module code. I'm not sure I'm a fan of using using statements because it makes it harder to copy paste parts into the terminal for testing but at least it's the same in all public functions now.

yea im a bit split on them as well, i sometimes use namespaces to reduce code width..
it does look make it cleaner, but can certainly make it harder to follow along.

> @trackd I hope I didn't break your tests/code switching them to the testconsole. It's part of the spectre.console.testing library you linked a while ago. The testconsole is pretty good, you can set it up and specify things like the width, features like ansi etc. so the tests are easier to get to work locally + in gh actions
> 

ah nice! I'm sure i can adapt the code to handle the new test method if needed.

> features like ansi
does that mean that we can enable the 24bit color tests in `ConvertTo-SpectreDecoration.tests.ps1` ?
(ie -Tag "ExcludeCI")

> And thanks for teaching me about that argtransformation attribute btw

It is pretty cool, only learned about it recently myself as well.
i've learned alot from your code and contributing to the module, it's alot of fun.


I'm not sure i understand the need for all the saving to .txt files or `-AsByteStream`
is file encoding messing something up so this is to normalize it?

```ps1
$snapShotComparison = "$PSScriptRoot\..\@snapshots\Format-SpectreBarChart.snapshot.compare.txt"
Set-Content -Path $snapShotComparison -Value ($testConsole.Output -replace "`r", "") -NoNewline
$compare = Get-Content -Path $snapShotComparison -AsByteStream
$snapshot = Get-Content -Path "$PSScriptRoot\..\@snapshots\Format-SpectreBarChart.snapshot.txt" -AsByteStream
$snapshot | Should -Be $compare
```

Importing the file with `-Raw` returns it as a single string object which should be comparable to `$testConsole.Output`

image


havnt finished going over all the changes, but looks very nice.

I have been thinking a bit, i know you split the module from a single .psm1 to 1 file per function.
That makes alot of sense for development.

But for distribution its usually preferred to use a single .psm1 as it's a bit faster to load.

this would remove all the using completers.psm1 statements as the class would be in the same file.
I've been looking at some buildtools [list](https://github.com/stars/trackd/lists/module-build-tools) (that's not a complete list, just a few i could find that seems active)

This would also enable building a dll from the parser C# code and adding it to the .psd1 (at buildtime)

`ModuleBuilder` is pretty popular.
i could take a look at it, if it's something you're interested in.

**Comment by trackd** [comment]
oh slightly off topic, but not sure if you've seen it, i use this a ton when trying to figure out the spectre lib methods and classes.

https://github.com/SeeminglyScience/ClassExplorer

Super useful module for poking around dotnet classes

File: PwshSpectreConsole/public/demo/Start-SpectreDemo.ps1 [hunk 1]
```diff
@@ -321,7 +322,7 @@ Invoke-SpectreCommandWithProgress -ScriptBlock {
 
     $example = @"
 Get-SpectreImageExperimental "$PSScriptRoot\..\..\private\images\harveyspecter.gif" -LoopCount 2
-Write-SpectreHost "I'm Harvey Specter. Are you after a Specter consult or a .?"
+Write-SpectreHost "I'm Harvey Specter. Are you after a Specter consult or a Spectre.Console?"
```
Comment (trackd) [inline PwshSpectreConsole/public/demo/Start-SpectreDemo.ps1:325]: maybe unintended regex replace? ;-)

File: PwshSpectreConsole.Tests/writing/Write-SpectreCalender.tests.ps1 [hunk 2]
```diff
@@ -68,5 +61,29 @@ Describe "Write-SpectreCalendar" {
...
+        It "Should match the snapshot" {
+            Mock Write-AnsiConsole {
+                $testConsole.Write($RenderableObject)
+            }
+            $events = @{
+                '2022-03-10' = 'Event 1'
+                '2022-03-20' = 'Event 2'
+            }
+            $culture = Get-Culture -Name "en-US"
+            Write-SpectreCalendar -Date 2024-07-01 -Culture $culture -Events $events -Border "Rounded" -Color "SpringGreen3"
```
Comment (trackd) [inline PwshSpectreConsole.Tests/writing/Write-SpectreCalender.tests.ps1:74]: -Culture is [cultureinfo] type,
so you can just pass 'en-us' to it dont need to do Get-Culture.

File: PwshSpectreConsole/PwshSpectreConsole.psd1 [hunk 3]
```diff
@@ -58,7 +58,8 @@ RequiredAssemblies =
                '.\packages\Spectre.Console\lib\netstandard2.0\Spectre.Console.dll', 
                '.\packages\Spectre.Console.ImageSharp\lib\netstandard2.0\Spectre.Console.ImageSharp.dll', 
                '.\packages\SixLabors.ImageSharp\lib\netstandard2.0\SixLabors.ImageSharp.dll', 
-               '.\packages\Spectre.Console.Json\lib\netstandard2.0\Spectre.Console.Json.dll'
+               '.\packages\Spectre.Console.Json\lib\netstandard2.0\Spectre.Console.Json.dll',
+               '.\packages\Spectre.Console.Testing\lib\netstandard2.0\Spectre.Console.Testing.dll'
```
Comment (trackd) [inline PwshSpectreConsole/PwshSpectreConsole.psd1:62]: Is that needed? is the testlib used in the public/private functions? 
i thought it was limited to pester tests which dont ship with the module

Comment (ShaunLawrie) [reply PwshSpectreConsole/PwshSpectreConsole.psd1:62]: I added it here to get it working for now. I think I'll add it with Add-Type in the @init test to import it unless there's a better way

**Comment by ShaunLawrie** [comment]
> Importing the file with -Raw returns it as a single string object which should be comparable to $testConsole.Output

As far as I could tell using -Raw still strips ansi which I need because ansi changes should probably be a failure until it can be explained. I will pull this snapshot logic out into the testhelpers because it's just copy pasted all over the place atm.

I like writing the file to the filesystem during snapshot testing because it makes it much easier to investigate after a test run. I will upload the snapshot and comparison files as github action artifacts if there are failures.

**Comment by trackd** [comment]
> > Importing the file with -Raw returns it as a single string object which should be comparable to $testConsole.Output
> 
> As far as I could tell using -Raw still strips ansi which I need because ansi changes should probably be a failure until it can be explained. I will pull this snapshot logic out into the testhelpers because it's just copy pasted all over the place atm.
> 

huh that is weird, Get-Content does not strip the colors for me, also works in 5.1 (with `-Encoding utf8`)
image

this seems to work for me, 
```ps1
# https://gist.github.com/trackd/d6d2bcae1fb0e7c8ba01e51153649ce5
Enable-DebugSpectre
ls | fst | Set-Content ..\table.spectre.txt -NoNewline
ls | fst | should -be (Get-Content -Raw ..\table.spectre.txt)
```

> I like writing the file to the filesystem during snapshot testing because it makes it much easier to investigate after a test run. I will upload the snapshot and comparison files as github action artifacts if there are failures.

i think that information is already in the error, i can usually see the diff when i test locally.
but it might truncate it a bit.. there is a ton of pester settings though.

i use this to run the tests, run in a new pwsh otherwise the `Enable-DebugSpectre` breaks the tests 😄 

```ps1
function Test-PwshSpectreConsole {
    <#
        https://pester.dev/docs/usage/test-results
        https://pester.dev/docs/usage/configuration
    #>
    [cmdletbinding()]
    param(
        $Name = "PwshSpectreConsole",
        $Folder = "D:\Code\PwshSpectreConsole\"
    )
    $runner = {
        param($Current, $Path)
        Write-Host 'Starting pwsh -noprofile and Running Pester tests...'
        Import-Module Pester -Force
        $Date = Get-Date -Format "yyyyMMdd_HHmm"
        $config = New-PesterConfiguration
        $config.Run.Path = $Path
        $config.CodeCoverage.Enabled = $true
        # $config.Debug.ShowFullErrors = $true
        # $config.Debug.WriteDebugMessages = $true
        $config.TestResult.OutputFormat = "NUnitXml"
        $config.TestResult.OutputPath = (Join-Path (Split-Path -Parent $Path) "test-results" "${Current}_${Date}_NUnit.xml")
        $config.TestResult.Enabled = $true
        $config.output.CIFormat = 'GithubActions'
        Invoke-Pester -Configuration $config
    }
    pwsh -NoProfile -NonInteractive -Command $runner -args $Name, $Folder
}

```

**Comment by ShaunLawrie** [comment]
> huh that is weird, Get-Content does not strip the colors for me, also works in 5.1 (with -Encoding utf8)
I don't know what I was doing to think that, -raw works fine.

**Pull request merged**